### ISSUE TITLE
Trace and span IDs in apiserver logs using contextual logging

### ISSUE THREAD

**Issue by dashpole** [kubernetes/kubernetes#119699]
### What would you like to be added?

Add `trace_id` and `span_id` keys to logs on the apiserver request path.  This can be done using the contextual logging feature: https://github.com/kubernetes/enhancements/issues/3077

cc @kubernetes/wg-structured-logging-leads @kubernetes/sig-instrumentation-leads @Richabanker 

### Why is this needed?

With https://github.com/kubernetes/enhancements/issues/647, we can now collect traces for requests to the kubernetes api.  After https://github.com/kubernetes/kubernetes/issues/119697, we will be able to link from metrics to trace and span IDs.  If we had those in logs, we would be able to link from metrics to logs, which would be useful.

**Comment by jiahuif** [comment]
/assign @Richabanker
Could you take a look? Thank you.
/triage accepted

**Comment by marefr** [comment]
Any updates on this and/or is it up for grabs?

**Comment by richabanker** [comment]
@marefr I haven't had the chance to take this up, feel free to do so! Thanks!

**Comment by marefr** [comment]
@richabanker thanks will do

**Comment by marefr** [comment]
I did find a workaround for my use case. Before open any PR though would be good with some quick feedback whether my idea is enough for this since it doesn't really involve contextual logging as mentioned in the title of this issue.

In [traces filter](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/filters/traces.go) suggests adding a new filter that adds key/value to the [httplog](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apiserver/pkg/server/httplog):
```go
// WithTracingHTTPLoggingAttributes adds tracing attributes to HTTP request logs.
func WithTracingHTTPLoggingAttributes(handler http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		spanCtx := trace.SpanContextFromContext(r.Context())
		if spanCtx.IsValid() {
			if spanCtx.HasTraceID() {
				httplog.AddKeyValue(r.Context(), "traceID", spanCtx.TraceID().String())
			}

			if spanCtx.HasSpanID() {
				httplog.AddKeyValue(r.Context(), "spanID", spanCtx.SpanID().String())
			}
		}

		handler.ServeHTTP(w, r)
	})
}
```

Then before [adding the http logging filter](https://github.com/kubernetes/kubernetes/blob/74d578485dcf076fdbbf3c09e1b0ae43c52c19d7/staging/src/k8s.io/apiserver/pkg/server/config.go#L1041), add above filter:
```go
if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.APIServerTracing) {
	handler = genericapifilters.WithTracingHTTPLoggingAttributes(handler)
}
```

It's added before http logging filter because it's executed in the reverse order and by that both a trace has been started and http logging initialized in the request context:
1. tracing
2. http logging
3. add traceID and spanID key/value's to http logging

Does this make sense or did you have anything else in mind?

**Comment by pohly** [comment]
I think the goal is to add `trace_id` and `span_id` to all log output along the path of handling a request, not just to "http logging" (the logging of the request/response?).

**Comment by dashpole** [comment]
I imagined it would be integrated with klog in some fashion, given we support contextual logging: https://github.com/kubernetes/klog/blob/ab53041d62fd518fc9f69ea518f6abed3bcd4905/contextual.go#L156

@pohly do you have any suggestions for the best way to do that?

**Comment by marefr** [comment]
Thanks. Then this might be to big/unclear for me to pick up. Looking into the klog contextual logger looks like you you can only create a contextual logger once so are you imagine that all code eventually gets a logger from context (if HTTP request there's a contextual logger created at beginning of request) and any code involved in handling the request uses a contextual logger for logging?

**Comment by pohly** [comment]
The solution would be somewhere along the lines of extracting `trace_id` and `span_id`, then adding them to a logger and that logger to the context:
```
logger := klog.FromContext(ctx)
logger = klog.LoggerWithValues(logger, "trace_id", traceID, "span_id", spanID)
ctx := klog.NewContext(ctx, logger)
doSomething(ctx, ...)
```

Beware that this has performance implications, so that would have to be evaluated.

**Comment by dashpole** [comment]
@marefr if you need to store any utility code (e.g. the code that wraps a logger and adds trace_id + span_id), you can place it in a directory here: https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/component-base/logs

**Comment by marefr** [comment]
Thanks all, that's helpful

> The other half of the problem is that code running inside `doSomething` must be converted to contextual logging. @tallclair wanted to do something in that area, but it only got to a RFC PR. We are tracking that in the description of [kubernetes/enhancements#3077](https://github.com/kubernetes/enhancements/issues/3077).

So [kubernetes/enhancements#3077](https://github.com/kubernetes/enhancements/issues/3077) is tracking converting to use contextual logger and doing part of that related to HTTP request logs here would be too much work/big PR. To limit the scope/size of PR, could a viable first step be the following? 

1. Setup a contextual logger with `trace_id` and `span_id` as suggested above. I imagine perhaps as part of the existing  [tracing filter](https://github.com/kubernetes/kubernetes/blob/84a7f1ea15e4fcabe88e06ca27de611fa3b74437/staging/src/k8s.io/apiserver/pkg/endpoints/filters/traces.go#L31) since it's already guarded with tracing feature flag check.
2. Change the http logging package to use a contextual logger 

When that's done what's left is basically continue converting to use contextual loggers everywhere, right? Part of that is  related to this issue so we can keep this open until [kubernetes/enhancements#3077](https://github.com/kubernetes/enhancements/issues/3077) is closed or close this and continue as is with [kubernetes/enhancements#3077](https://github.com/kubernetes/enhancements/issues/3077).

Thoughts?

> Beware that this has performance implications, so that would have to be evaluated.

Can you share any details why that is?

**Comment by pohly** [comment]
Conversion to contextual logging is incremental. We try to do it per-package, but picking a smaller subset also makes sense.

Performance is affected because all of those calls need to do some amount of allocations, which can matter if `doSomething` is very cheap.

**Comment by marefr** [comment]
> Performance is affected because all of those calls need to do some amount of allocations, which can matter if `doSomething` is very cheap.

Got it. Out of curiosity you've already agreed/working on converting to use contextual logging so haven't you already done some performance benchmarks?

**Comment by pohly** [comment]
I did benchmarks for the scheduler and micro-benchmarks for the code. The `klog.FromContext` call is fast enough that converting code to use it doesn't matter.

The `WithValues` and `NewContext` calls are more expensive. We had to limit their usage in the scheduler to higher log levels to avoid performance regressions at "production" verbosity levels. For them, the code owners need to decide how much overhead for better log output they are willing to accept.

**Comment by rexagod** [comment]
@marefr Are you still planning to work on this?

**Comment by marefr** [comment]
@rexagod no

**Comment by rexagod** [comment]
@pohly ~Do we plan on working on this in the future?~

IIUC This seems to be blocked by https://github.com/kubernetes/enhancements/issues/3077.

**Comment by pohly** [comment]
This is independent from contextual logging.We have the means now, someone would need to figure out how to apply them. In other words, this is waiting for a volunteer...

**Comment by marefr** [comment]
@pohly does that mean that my earlier suggestion now is viable way forward? If not, I think you have the knowledge and details to explain the meaning further so that a volunteer can pick it up.

> To limit the scope/size of PR, could a viable first step be the following?
> 
> 1. Setup a contextual logger with `trace_id` and `span_id` as suggested above. I imagine perhaps as part of the existing  [tracing filter](https://github.com/kubernetes/kubernetes/blob/84a7f1ea15e4fcabe88e06ca27de611fa3b74437/staging/src/k8s.io/apiserver/pkg/endpoints/filters/traces.go#L31) since it's already guarded with tracing feature flag check.
> 2. Change the http logging package to use a contextual logger

**Comment by pohly** [comment]
Sorry, I don't have time to dig into the details at the moment.
